// @globalhive/vuejs-tour@1.6.0/dist/vuejs-tour.js
import { ref as b, reactive as B, computed as p, onMounted as E, openBlock as j, createElementBlock as T, Fragment as N, createElementVNode as n, renderSlot as _, normalizeProps as L, guardReactiveProps as w, withModifiers as g, toDisplayString as m, createCommentVNode as P } from "vue";
import { createPopper as A } from "@popperjs/core";
import k from "jump.js";
const M = /* @__PURE__ */ n("div", {
  id: "vjt-backdrop",
  "data-hidden": ""
}, null, -1), V = {
  id: "vjt-tooltip",
  "data-hidden": "",
  role: "tooltip"
}, D = ["innerHTML"], F = { class: "vjt-actions" }, H = ["textContent"], O = ["textContent"], z = ["textContent"], R = /* @__PURE__ */ n("div", {
  id: "vjt-arrow",
  "data-popper-arrow": ""
}, null, -1), G = {
  __name: "VTour",
  props: {
    name: {
      type: String,
      default: "default"
    },
    steps: {
      type: Array,
      required: !0
    },
    backdrop: {
      type: Boolean,
      default: !1
    },
    autoStart: {
      type: Boolean,
      default: !1
    },
    startDelay: {
      type: Number,
      default: 0
    },
    highlight: {
      type: Boolean,
      default: !1
    },
    buttonLabels: {
      type: Object,
      default: () => ({
        next: "Next",
        prev: "Back",
        finish: "Finish",
        skip: "Skip"
      })
    },
    saveToLocalStorage: {
      type: String,
      default: "end"
    }
  },
  emits: ["onTourStart", "onTourEnd"],
  setup(c, { expose: v, emit: x }) {
    const f = b(!1), o = b(null), t = B({
      currentStep: 0,
      lastStep: null,
      getCurrentStep: p(() => e.steps[t.currentStep]),
      getLastStep: p(() => e.steps[t.lastStep])
    }), s = p(() => e.steps.length - 1), I = p(() => t.currentStep === s.value ? e.buttonLabels.finish : e.buttonLabels.next), e = c, h = x;
    v({
      startTour: a,
      nextStep: i,
      prevStep: S,
      endTour: u,
      goToStep: $,
      resetTour: q
    });
    async function a() {
      switch (e.saveToLocalStorage) {
        case "end":
          if (localStorage.getItem("vjt-" + e.name) === "true")
            return;
          break;
        case "step":
          if (localStorage.getItem("vjt-" + e.name) === "true")
            return;
          if (localStorage.getItem("vjt-" + e.name)) {
            t.currentStep = localStorage.getItem("vjt-" + e.name), t.lastStep = localStorage.getItem("vjt-" + e.name) - 1 >= 0 ? localStorage.getItem("vjt-" + e.name) - 1 : 0;
            break;
          }
          break;
      }
      if (document.querySelector(`${t.getCurrentStep.target}`) === null) {
        t.currentStep++, a();
        return;
      }
      setTimeout(() => {
        e.backdrop === !0 && document.getElementById("vjt-backdrop").removeAttribute("data-hidden"), document.getElementById("vjt-tooltip").removeAttribute("data-hidden"), o.value = A(document.querySelector(`${t.getCurrentStep.target}`), document.getElementById("vjt-tooltip"), {
          placement: `${t.getCurrentStep.placement ? t.getCurrentStep.placement : "top"}`,
          modifiers: [
            {
              name: "offset",
              options: {
                offset: [0, 8]
              }
            }
          ]
        }), e.highlight && y(), h("onTourStart"), f.value = !0;
      }, e.startDelay);
    }
    function y() {
      var C;
      let r = document.querySelector(`${t.getCurrentStep.target}`), l = document.querySelector(`${(C = t.getLastStep) == null ? void 0 : C.target}`);
      r.classList.add("vjt-highlight"), l != null && r !== l && l.classList.remove("vjt-highlight");
    }
    async function i() {
      if (t.currentStep < s.value) {
        for (t.getCurrentStep.onNext && await t.getCurrentStep.onNext(), t.lastStep = t.currentStep, t.currentStep++, e.saveToLocalStorage === "step" && localStorage.setItem("vjt-" + e.name, t.currentStep); document.querySelector(`${t.getCurrentStep.target}`) === null; ) {
          if (t.currentStep <= s.value)
            return u();
          t.currentStep++;
        }
        await d();
        return;
      }
      u();
    }
    async function S() {
      if (t.currentStep > 0) {
        for (t.getCurrentStep.onPrev && await t.getCurrentStep.onPrev(), t.lastStep = t.currentStep, t.currentStep--, e.saveToLocalStorage === "step" && localStorage.setItem("vjt-" + e.name, t.currentStep); document.querySelector(`${t.getCurrentStep.target}`) === null; )
          t.currentStep--;
        await d();
      }
    }
    function u() {
      var r;
      document.getElementById("vjt-tooltip").setAttribute("data-hidden", ""), document.getElementById("vjt-backdrop").setAttribute("data-hidden", ""), (r = document.querySelector(".vjt-highlight")) == null || r.classList.remove("vjt-highlight"), o.value.destroy(), (e.saveToLocalStorage === "end" || e.saveToLocalStorage === "step") && localStorage.setItem("vjt-" + e.name, "true"), k(document.body, {
        duration: 500
      }), h("onTourEnd");
    }
    async function $(r) {
      for (f.value === !1 && await a(), t.getCurrentStep.onNext && await t.getCurrentStep.onNext(), t.lastStep = r - 1 >= 0 ? r - 1 : 0, t.currentStep = r, e.saveToLocalStorage === "step" && localStorage.setItem("vjt-" + e.name, t.currentStep); document.querySelector(`${t.getCurrentStep.target}`) === null; )
        t.currentStep++;
      await d();
    }
    function q() {
      t.currentStep = 0, t.lastStep = 0, (e.saveToLocalStorage === "end" || e.saveToLocalStorage === "step") && localStorage.setItem("vjt-" + e.name, "true"), a();
    }
    async function d() {
      await o.value.setOptions({
        placement: `${t.getCurrentStep.placement ? t.getCurrentStep.placement : "top"}`
      }), o.value.state.elements.reference = document.querySelector(`${t.getCurrentStep.target}`), await o.value.update(), k(document.querySelector(`${t.getCurrentStep.target}`), {
        duration: 500,
        offset: -100
      }), e.highlight && y(), t.getCurrentStep.onShow && await t.getCurrentStep.onShow();
    }
    return E(() => {
      e.autoStart && a();
    }), (r, l) => (j(), T(N, null, [
      M,
      n("div", V, [
        _(r.$slots, "content", L(w({ step: t })), () => [
          n("div", {
            innerHTML: t.getCurrentStep.content
          }, null, 8, D)
        ]),
        _(r.$slots, "actions", L(w({ endTour: u, nextStep: i, prevStep: S, step: t })), () => [
          n("div", F, [
            n("button", {
              type: "button",
              onClick: g(u, ["prevent"]),
              textContent: m(e.buttonLabels.skip)
            }, null, 8, H),
            t.currentStep > 0 ? (j(), T("button", {
              key: 0,
              type: "button",
              textContent: m(e.buttonLabels.prev),
              onClick: g(S, ["prevent"])
            }, null, 8, O)) : P("", !0),
            n("button", {
              type: "button",
              textContent: m(I.value),
              onClick: g(i, ["prevent"])
            }, null, 8, z)
          ])
        ]),
        R
      ])
    ], 64));
  }
}, U = {
  install: (c, v) => {
    c.component("VTour", G);
  }
};
export {
  U as default
};
